### ISSUE TITLE
[QUESTION]gq8b1uv dimmer throws an error while trying to adjust brightness

### ISSUE THREAD

**Issue by peca89** [xaviml/controllerx#69]
# Question

Hello. I'm not sure if problem is up to ControllerX, HA or Z2M and I don't know how to find out, so I'm starting from here. The problem is controlling the brightness level of [gq8b1uv](https://www.zigbee2mqtt.io/devices/gq8b1uv.html) dimmer, while the simple turn on/off works fine. I'm trying to control it's brightness using IKEA E1524's color buttons (and therefore using `CustomLightController`) because I want brightness buttons to control other Tradfri bulb. The controllerX throws an error, as given below. The exact same config works if I just replace the `light` attribute with Tradfri bulb.

Please try to point me in the right direction. 

## Additional information
Using latest versions of everything as of today
## AppDaemon app configuration

```yaml
livingroom_controller3:
  module: controllerx
  class: CustomLightController
  controller: sensor.0x90fd9ffffef5c2dd_action
  integration: z2m
  light: light.0xec1bbdfffec3e6d1_light
  mapping:
    arrow_right_hold: hold_brightness_up
    arrow_right_release: release
    arrow_left_hold: hold_brightness_down
    arrow_left_release: release
```

### AppDaemon
```text
2020-04-25 10:29:36.242690 INFO livingroom_controller3: 🎮 Button event triggered: `arrow_left_hold`
2020-04-25 10:29:36.251647 INFO livingroom_controller3: 🏃 Running `hold_brightness_down` now
2020-04-25 10:29:36.284769 WARNING livingroom_controller3: ------------------------------------------------------------
2020-04-25 10:29:36.286975 WARNING livingroom_controller3: Unexpected error in worker for App livingroom_controller3:
2020-04-25 10:29:36.289458 WARNING livingroom_controller3: Worker Ags: {'id': '41e5a0fb0e0f476bb4415e62d3bf588f', 'name': 'livingroom_controller3', 'objectid': '01625a3a198c4a96bd0d13e9d1fc10ca', 'type': 'state', 'function': <bound method StateIntegration.callback of <core.integration.z2m.Z2MIntegration object at 0x7f296587a0a0>>, 'attribute': 'state', 'entity': 'sensor.0x90fd9ffffef5c2dd_action', 'new_state': 'arrow_left_hold', 'old_state': '', 'pin_app': True, 'pin_thread': 3, 'kwargs': {'__thread_id': 'MainThread'}}
2020-04-25 10:29:36.291992 WARNING livingroom_controller3: ------------------------------------------------------------
2020-04-25 10:29:36.295965 WARNING livingroom_controller3: Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/appdaemon/threading.py", line 709, in async_worker
    await funcref(entity, attr, old_state, new_state, self.AD.state.sanitize_state_kwargs(app, args["kwargs"]))
  File "/config/appdaemon/apps/controllerx/core/integration/state.py", line 15, in callback
    await self.controller.handle_action(new)
  File "/config/appdaemon/apps/controllerx/core/controller.py", line 141, in handle_action
    await self.call_action(action_key)
  File "/config/appdaemon/apps/controllerx/core/controller.py", line 159, in call_action
    await self.action_timer_callback({"action_key": action_key})
  File "/config/appdaemon/apps/controllerx/core/controller.py", line 170, in action_timer_callback
    await action(*args)
  File "/config/appdaemon/apps/controllerx/core/controller.py", line 31, in _action_impl
    await method(self, *args, **kwargs)
  File "/config/appdaemon/apps/controllerx/core/type/light_controller.py", line 383, in hold
    await super().hold(attribute, direction)
  File "/config/appdaemon/apps/controllerx/core/controller.py", line 31, in _action_impl
    await method(self, *args, **kwargs)
  File "/config/appdaemon/apps/controllerx/core/controller.py", line 260, in hold
    stop = await self.hold_loop(*args)
  File "/config/appdaemon/apps/controllerx/core/type/light_controller.py", line 386, in hold_loop
    return await self.change_light_state(
  File "/config/appdaemon/apps/controllerx/core/type/light_controller.py", line 419, in change_light_state
    new_state_attribute, exceeded = stepper.step(old, direction)
  File "/config/appdaemon/apps/controllerx/core/stepper/minmax_stepper.py", line 21, in step
    new_value = value + sign * step
TypeError: can only concatenate str (not "float") to str

2020-04-25 10:29:36.301402 WARNING livingroom_controller3: ------------------------------------------------------------
2020-04-25 10:29:40.550613 INFO livingroom_controller3: 🎮 Button event triggered: `arrow_left_release`
2020-04-25 10:29:40.558527 INFO livingroom_controller3: 🏃 Running `release` now
```
### Z2M
Firstly, light is turned on using `brightness_down_click` and then I try to turn the brightness down using `arrow_left_hold`
```text
zigbee2mqtt:info  2020-04-25 10:28:50: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd/action', payload 'brightness_down_click'
zigbee2mqtt:info  2020-04-25 10:28:50: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":55,"battery":47,"action":"brightness_down_click"}'
zigbee2mqtt:info  2020-04-25 10:28:50: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":55,"battery":47,"action":""}'
zigbee2mqtt:info  2020-04-25 10:28:53: MQTT publish: topic 'zigbee2mqtt/0xec1bbdfffec3e6d1', payload '{"state":"ON","linkquality":68,"brightness":"254.00","level":1000}'
zigbee2mqtt:info  2020-04-25 10:29:49: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":47,"battery":34}'
zigbee2mqtt:info  2020-04-25 10:29:49: MQTT publish: topic 'homeassistant/device_automation/0x90fd9ffffef5c2dd/action_arrow_left_hold/config', payload '{"automation_type":"trigger","type":"action","subtype":"arrow_left_hold","payload":"arrow_left_hold","topic":"zigbee2mqtt/0x90fd9ffffef5c2dd/action","device":{"identifiers":["zigbee2mqtt_0x90fd9ffffef5c2dd"],"name":"0x90fd9ffffef5c2dd","sw_version":"Zigbee2mqtt 1.12.2","model":"TRADFRI remote control (E1524/E1810)","manufacturer":"IKEA"}}'
zigbee2mqtt:info  2020-04-25 10:29:49: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":47,"battery":34,"action":"arrow_left_hold"}'
zigbee2mqtt:info  2020-04-25 10:29:49: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":47,"battery":34,"action":""}'
zigbee2mqtt:info  2020-04-25 10:29:49: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd/action', payload 'arrow_left_hold'
zigbee2mqtt:info  2020-04-25 10:29:53: MQTT publish: topic 'homeassistant/device_automation/0x90fd9ffffef5c2dd/action_arrow_left_release/config', payload '{"automation_type":"trigger","type":"action","subtype":"arrow_left_release","payload":"arrow_left_release","topic":"zigbee2mqtt/0x90fd9ffffef5c2dd/action","device":{"identifiers":["zigbee2mqtt_0x90fd9ffffef5c2dd"],"name":"0x90fd9ffffef5c2dd","sw_version":"Zigbee2mqtt 1.12.2","model":"TRADFRI remote control (E1524/E1810)","manufacturer":"IKEA"}}'
zigbee2mqtt:info  2020-04-25 10:29:53: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":44,"battery":34,"action":"arrow_left_release","duration":4.78}'
zigbee2mqtt:info  2020-04-25 10:29:54: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd', payload '{"linkquality":44,"battery":34,"action":""}'
zigbee2mqtt:info  2020-04-25 10:29:54: MQTT publish: topic 'zigbee2mqtt/0x90fd9ffffef5c2dd/action', payload 'arrow_left_release'
```

**Comment by xaviml** [comment]
Hi @peca89,

You say that the same configuration works with another Tradfri light, right? but it does not work with `light.0xec1bbdfffec3e6d1_light`. Which light is this one, also a Tradfri? How is the light integrated? z2m as well?

The error you see in ControllerX is because the brightness value could not be read from the state attributes of the light `light.0xec1bbdfffec3e6d1_light`. Could you maybe check if `brightness` attribute is in `attributes` column of the entity `light.0xec1bbdfffec3e6d1_light` from "Developer Tools > States"? Try to turn on and off the light to see if the `brightness` disappears when toggling the light.

If you answer these questions, I might be able to help you further with your problem, but by experience, that error should not appear in ControllerX and if it does it is because the light does not support the attribute. I have also changed the code so that error should not appear anymore and you will see a much clearer message.

Could you also please send me the `attributes` column from the light `light.0xec1bbdfffec3e6d1_light` when the light is off and on?

Thank you :)

**Comment by peca89** [comment]
> You say that the same configuration works with another Tradfri light, right? but it does not work with `light.0xec1bbdfffec3e6d1_light`. Which light is this one, also a Tradfri? How is the light integrated? z2m as well?

Yes, the sam config works with a Tradfri light which is not mentioned in the logs above. The `light.0xec1bbdfffec3e6d1_light` is an in-wall dimmer switch, produced by [Bilisty](https://www.alibaba.com/product-detail/zigbee-dimmer-in-wall-switch-light_62510813273.html?spm=a2700.icbuShop.41413.21.63034c52M6LeK3&fullFirstScreen=true) which is identified as gq8b1uv because it uses the Tuya zigbee module inside. It uses z2m. 

> The error you see in ControllerX is because the brightness value could not be read from the state attributes of the light `light.0xec1bbdfffec3e6d1_light`. Could you maybe check if `brightness` attribute is in `attributes` column of the entity `light.0xec1bbdfffec3e6d1_light` from "Developer Tools > States"? Try to turn on and off the light to see if the `brightness` disappears when toggling the light.

Now that's a pointing in the right direction! :) It seems that I have found a problem. It seems that the brightness value is reported as a string:

```yaml
brightness: '254.00'    <-----QUOTES ?!?!?
state: 'ON'
linkquality: 68
level: 1000
friendly_name: Dimmer_DS_light
supported_features: 41
```

If you look at my z2m log from above, it is also reported as string. 
`zigbee2mqtt:info  2020-04-25 10:28:53: MQTT publish: topic 'zigbee2mqtt/0xec1bbdfffec3e6d1', payload '{"state":"ON","linkquality":68,"brightness":"254.00","level":1000}'`
I'll try to dig into z2m files to hunt for the bug. Many thanks

**Comment by xaviml** [comment]
Indeed, you spotted the error. The error you see in ControllerX `TypeError: can only concatenate str (not "float") to str` is because the`value` variable is the `brightness` from the state attributes, which is a string. While you dig into the problem in z2m, I will see if I can patch this to convert string numbers into floats. This would fix your problem, although the actual problem resides in z2m. I hope this helps, I will keep you posted with the fix from ControllerX. Thanks!

**Comment by peca89** [comment]
OK, I've found it and fixed it. It's [here](https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/converters/fromZigbee.js#L4340) in a file `fromZigbee.js` within zigbee-herdsman-converters.
```javascript
    tuya_dimmer: {
        cluster: 'manuSpecificTuyaDimmer',
        type: 'commandGetData',
        convert: (model, msg, publish, options, meta) => {
            const key = msg.data.dp;
            const val = msg.data.data;
            if (key === 257) {
                return {state: (val[0]) ? 'ON': 'OFF'};
            } else {
                const level = val[2]*256 + val[3];
                const normalised = (level - 10) / (1000 - 10);
                return {brightness: (normalised * 254).toFixed(2), level: level};
            }
        },
    },
```
It seems that `toFixed()` function returns a string in JS. Anyway, I have replaced that with 
`Math.round()` and now it works fine. Integer brightness is precise enough.

Anyway, this is topic for z2m. I'll suggest a modification there and let someone more experienced review if that breaks anything else.

Thanks for your time.

**Comment by xaviml** [comment]
Hi @peca89,

Thank you for letting me know. Apparently, before you mentioned this error, I have controlled it without realizing it. I have been refactoring code to make it more robust by adding typing annotations and check the styling with flake8 and this changed the function that gets the value for brightness and color temp to then converting them to float.

On the next release of ControllerX, if the brightness or color temp are string, they will be converted to float. However, it is good that this gets solved from z2m as you mentioned.

**Comment by xaviml** [comment]
Although this is fixed on z2m, I added the fix in ControllerX, so it should not happen again. This is now released in v2.5.0.

**Comment by epicurean70** [comment]
Hi, Can someone share where I can buy this zigbee dimmer?
